ui/src/components/friends/add.rs

```diff
@@ -8,32 +12,108 @@ use kit::{
-        replace_spaces_underscore: false,
-        // Show a clear button inside the input field
-        with_clear_btn: true,
-        // Use the default options for the remaining fields
-        ..Options::default()
-    };
+    // todo: add translations for toasts
+    if *request_sent.get() {
+        state
+            .write()
+            .mutate(Action::AddToastNotification(ToastNotification::init(
+                "".into(),
+                "Friend Request Sent!".into(),
+                None,
+                5,
+            )));
+        request_sent.set(false);
+    }
+
+    if let Some(id) = my_id.get().clone() {
+        let mut clipboard = Clipboard::new().unwrap();
+        clipboard.set_text(id).unwrap();
+        state
+            .write()
+            .mutate(Action::AddToastNotification(ToastNotification::init(
+                "".into(),
+                "Copied ID to clipboard!".into(),
+                None,
+                5,
+            )));
+        my_id.set(None);
+    }
+
+    let ch = use_coroutine(cx, |mut rx: UnboundedReceiver<DID>| {
+        to_owned![request_sent];
+        async move {
+            let warp_cmd_tx = WARP_CMD_CH.tx.clone();
+            while let Some(did) = rx.next().await {
+                let (tx, rx) = oneshot::channel::<Result<(), warp::error::Error>>();
+                warp_cmd_tx
+                    .send(WarpCmd::MultiPass(MultiPassCmd::RequestFriend {
+                        did,
+                        rsp: tx,
+                    }))
+                    .expect("AddFriendLayout failed to send warp command");
+
+                let res = rx.await.expect("failed to get response from warp_runner");
+                match res {
+                    Ok(_) => request_sent.set(true),
+                    Err(e) => match e {
+                        Error::CannotSendSelfFriendRequest
+                        | Error::IdentityDoesntExist
+                        | Error::BlockedByUser
+                        | Error::InvalidIdentifierCondition => {
+                            // todo: show an error message
+                        }
+                        _ => todo!("failed to send friend request"),
+                    },
+                }
+            }
+        }
+    });
+
+    let id_ch = use_coroutine(cx, |mut rx: UnboundedReceiver<()>| {
+        to_owned![my_id];
+        async move {
+            let warp_cmd_tx = WARP_CMD_CH.tx.clone();
+            while rx.next().await.is_some() {
+                let (tx, rx) = oneshot::channel::<Result<DID, warp::error::Error>>();
+                warp_cmd_tx
+                    .send(WarpCmd::MultiPass(MultiPassCmd::GetOwnIdentity { rsp: tx }))
+                    .expect("AddFriendLayout failed to send warp command");
+
+                let res = rx.await.expect("failed to get response from warp_runner");
+                match res {
+                    Ok(did) => my_id.set(Some(did.to_string())),
```

Comment:
this goes to the clipboard

from line 55
```
  if let Some(id) = my_id.get().clone() {
        let mut clipboard = Clipboard::new().unwrap();
        clipboard.set_text(id).unwrap();
        state
            .write()
            .mutate(Action::AddToastNotification(ToastNotification::init(
                "".into(),
                "Copied ID to clipboard!".into(),
                None,
                5,
            )));
        my_id.set(None);
    }
```

Reply:
Okay thats fine then, though it might be best long term to pass the DID or Identity and grab it from there (especially if you will be displaying your own identity in another component)